### DISCUSSION TITLE
Fill Components with HashMap Fields using Scenes

### DISCUSSION THREAD

**Discussion by pubrrr** [bevyengine/bevy#3609] [Q&A] [answered]
I wanted to use scenes to create menus, different levels, etc. However I encountered the following problem:

For starters, I wanted to generate scenes by using a dummy world, similar to the [scenes example](https://github.com/bevyengine/bevy/blob/v0.6.0/examples/scene/scene.rs#L80-L95). 

- use a dummy world
- insert the components with the desired values
- use the world to generate a dynamic scene
- spawn this scene

From compiler errors relating to reflection, I figured that I have to use `bevy::utils::HashMap` instead of `std::collections::HashMap`. The components then are there, but **their `bevy::utils::HashMap`-fields are not correctly filled**. Is this a bug? Am I doing something wrong? Can someone please help me here?

Below is a code snipped where I reproduced the issue in a minimal setup - using Bevy 0.6:
(The example also shows that primitives and `Vec`s work fine)

```
use bevy::core::FixedTimestep;
use bevy::DefaultPlugins;
use bevy::prelude::{App, Assets, Component, DynamicScene, Query, Reflect, ReflectComponent, Res, ResMut, SceneSpawner, SystemSet, World};
use bevy::reflect::TypeRegistryArc;
use bevy::utils::HashMap;

fn main() {
    App::new()
        .register_type::<MyComponent>()
        .add_plugins(DefaultPlugins)
        .add_startup_system(startup)
        .add_system_set(
            SystemSet::new()
                .with_run_criteria(FixedTimestep::step(1.))
                .with_system(print_component),
        )
        .run();
}

#[derive(Component, Reflect, Debug, Default)]
#[reflect(Component)]
struct MyComponent {
    field: u32,
    vec: Vec<u8>,
    map: HashMap<u32, String>,
}

fn startup(
    mut scene_spawner: ResMut<SceneSpawner>,
    mut scenes: ResMut<Assets<DynamicScene>>,
    type_registry: Res<TypeRegistryArc>,
) {
    let mut world = World::new();

    let mut map = HashMap::default();
    map.insert(5, String::from("5"));
    map.insert(6, String::from("6"));
    world.spawn().insert(MyComponent { field: 5, vec: vec![7], map });

    let scene = DynamicScene::from_world(&world, &type_registry);
    let handle = scenes.add(scene);
    scene_spawner.spawn_dynamic(handle);
}

fn print_component(query: Query<&MyComponent>) {
    // continues to print:
    // s = MyComponent {
    //     field: 5,
    //     vec: [
    //         7,
    //     ],
    //     map: {},
    // }
    // --
    
    query.iter()
        .for_each(|s| eprintln!("s = {:#?}", s));
    println!("--");
}
```

**Comment by Davier** [comment, marked as answer]
The current behaviour when loading a map is to only apply modifications to existing keys, not inserting new ones. #1395 was merged recently, so we should now be able to properly insert new keys.
I think it's worth opening an issue about it so we don't forget, but it should be an easy PR :smile:

**Reply by pubrrr** [reply]
Thanks for the reply. I didn't have time to read the full PR yet. Does that mean it might be enough to add `#[derive(FromReflect)]` to my struct?

**Reply by Davier** [reply]
It will need some change in bevy itself, what I meant is that it should be an easy fix since the preliminary work has been done. I might be able to do it this weekend.

**Reply by Davier** [reply]
Done: #3701

**Reply by pubrrr** [reply]
Thank you :)

**Reply by lafi-99** [reply]
> The current behaviour when loading a map is to only apply modifications to existing keys, not inserting new ones. #1395 was merged recently, so we should now be able to properly insert new keys.
> 
> I think it's worth opening an issue about it so we don't forget, but it should be an easy PR :smile:
> 
>

**Reply by lafi-99** [reply]
> Thank you :)

**Reply by lafi-99** [reply]
> > The current behaviour when loading a map is to only apply modifications to existing keys, not inserting new ones. #1395 was merged recently, so we should now be able to properly insert new keys.
> > 
> > I think it's worth opening an issue about it so we don't forget, but it should be an easy PR :smile:
> > 
> > 
> 
>
